rasterio/stack.py

```diff
@@ -0,0 +1,362 @@
+        the case of file output.
+    nodata: float, optional
+        nodata value to use in output file. If not set, uses the nodata
+        value in the first input raster.
+    dtype: numpy.dtype or string
+        dtype to use in outputfile. If not set, uses the dtype value in
+        the first input raster.
+    indexes : list of ints or a single int, optional
+        bands to read and stack.
+    output_count: int, optional
+        If using callable it may be useful to have additional bands in
+        the output in addition to the indexes specified for read.
+    resampling : Resampling, optional
+        Resampling algorithm used when reading input files.
+        Default: `Resampling.nearest`.
+    target_aligned_pixels : bool, optional
+        Whether to adjust output image bounds so that pixel coordinates
+        are integer multiples of pixel size, matching the ``-tap``
+        options of GDAL utilities.  Default: False.
+    mem_limit : int, optional
+        Process stack output in chunks of mem_limit MB in size.
+    dst_path : str or PathLike, optional
+        Path of output dataset.
+    dst_kwds : dict, optional
+        Dictionary of creation options and other paramters that will be
+        overlaid on the profile of the output dataset.
+
+    Returns
+    -------
+    tuple
+        Two elements:
+            dest: numpy.ndarray
+                Contents of all input rasters in single array
+            out_transform: affine.Affine()
+                Information for mapping pixel coordinates in `dest` to
+                another coordinate system
+
+    Raises
+    ------
+    StackError
+        When sources cannot be stacked due to incompatibility between
+        them or limitations of the tool.
+    """
+    # Create a dataset_opener object to use in several places in this function.
+    if isinstance(sources[0], (str, os.PathLike)):
+        dataset_opener = rasterio.open
+    else:
+
+        @contextmanager
+        def nullcontext(obj):
+            try:
+                yield obj
+            finally:
+                pass
+
+        dataset_opener = nullcontext
+
+    dst = None
+
+    with ExitStack() as exit_stack:
+        with dataset_opener(sources[0]) as first:
+            first_profile = first.profile
+            first_crs = first.crs
+            best_res = first.res
+            first_nodataval = first.nodatavals[0]
+            nodataval = first_nodataval
+            dt = first.dtypes[0]
+
+            if indexes is None:
+                indexes = [None for s in sources]
+
+            try:
+                first_colormap = first.colormap(1)
+            except ValueError:
+                first_colormap = None
+
+        # scan input files
+        xs = []
+        ys = []
+        output_count = 0
+
+        for i, (dataset, src_indexes) in enumerate(zip(sources, indexes)):
+            with dataset_opener(dataset) as src:
```

Comment:
Would it make sense to open each dataset once (and store them in a list) and reuse them throughout the function instead of reopening them multiple times?